### ISSUE TITLE
Improve defaults for value labels border size

### ISSUE THREAD

**Issue by dej611** [elastic/elastic-charts#886]
As discussed in https://github.com/elastic/kibana/pull/81776 default for XY charts bar value labels border stylings can be improved moving to a "thicker" alternative.

### Current state

<img width="764" alt="Screenshot 2020-11-02 at 12 33 27" src="https://user-images.githubusercontent.com/924948/97899667-e3f8b200-1d39-11eb-8234-1f9d11bb6927.png">
<img width="756" alt="Screenshot 2020-11-02 at 12 38 26" src="https://user-images.githubusercontent.com/924948/97899727-ee1ab080-1d39-11eb-9584-41ff74454ab5.png">
<img width="762" alt="Screenshot 2020-11-02 at 12 47 10" src="https://user-images.githubusercontent.com/924948/97899747-f2df6480-1d39-11eb-850a-46dd667b6efc.png">

### Proposed new default

<img width="764" alt="Screenshot 2020-11-02 at 12 36 21" src="https://user-images.githubusercontent.com/924948/97899774-fe329000-1d39-11eb-9bfe-9c7e06caeb16.png">
<img width="792" alt="Screenshot 2020-11-02 at 12 37 07" src="https://user-images.githubusercontent.com/924948/97899791-038fda80-1d3a-11eb-94f9-4305d16aae71.png">
<img width="762" alt="Screenshot 2020-11-02 at 12 50 34" src="https://user-images.githubusercontent.com/924948/97899808-0ab6e880-1d3a-11eb-84f0-aab66b9e750d.png">

**Comment by markov00** [comment]
Approved for me. I agree that a thicker border makes it easier to read

**Comment by markov00** [comment]
cc @monfera @rshen91 @nickofthyme I'd like also your opinion on this.
Everything started from a Marco's PR about adding shadow/border to the text displayed on top of bars to increase the readability: https://github.com/elastic/elastic-charts/pull/785

**Comment by nickofthyme** [comment]
I concur 👍

**Comment by rshen91** [comment]
It definitely improves readability having a thicker border/shadow 👍

**Comment by ryankeairns** [comment]
On the surface, I see the contrast gained between the stroke and the text color but I've never researched this from an a11y standpoint (I've dropped a note into our internal Slack out of curiosity).

Additionally, there is an alternative EUI color palette for cases where overlaid text is needed. We should check on that approach as well before settling on the white outline. See the `euiPaletteColorBlindBehindText({ sortBy: 'natural' })` example on this page https://elastic.github.io/eui/#/utilities/color-palettes

cc:/ @cchaos

**Comment by markov00** [comment]
Thanks @ryankeairns 
from the point of view of elastic-charts that border is just another configurable option on the theme object.
The final stylistic approach should go into EUI theme created specifically for elastic-charts in Kibana and not in elastic-charts directly, our "defaults"  are not meant to be used in Kibana as they are.
cc @dej611

**Comment by markov00** [comment]
we should probably transfer that issue to EUI repo

**Comment by miukimiu** [comment]
I think this option makes sense. From an a11y standpoint, we should use the border to avoid situations like the following one:

<img width="771" alt="95337512-33ba9980-08b2-11eb-880b-66ec1b01426c" src="https://user-images.githubusercontent.com/2750668/98249806-99ac4680-1f6e-11eb-88e0-038805aceca1.png">
 
It's difficult to predict what is going to be the page background color. In this situation, the background is white and value white. Without the border, the readability of the `0.25` value would be very difficult. 

I'll open an issue in EUI so we can start thinking about improving the theme and make it a standalone package.

**Comment by ryankeairns** [comment]
+1 this approach seems generally safe from an a11y perspective:

> Text effects, like outlines, can impact perceived contrast. WCAG 2 states that the color of a text outline or border can be used as the text or foreground color when measuring contrast.

(source: https://webaim.org/articles/contrast/#outline)

**Comment by monfera** [comment]
I agree with the motive but the first two of the thicker, contrasty shadows under "Proposed new defaults" look busy or overpowering to me to the extent of lessening readability, while the third one looks about OK.

So I think that capping (limiting) the border width in terms of screen pixel or EM/REM would be better than scaling it with the font size. Moreover, the larger the apparent font, the less help it needs—and above some font size, it looks out of place altogether. Even with small font sizes, the extra white area doesn't automatically aid readability, because it only works as long as it helps our eyes detect edges, but too much border will establish a perceptual surface on its own, with its own separate edge detection of both edges of the border :-) 

So we shouldn't see it as contrast enhancer—that'd need larger surfaces—but merely as "edge maker/enhancer".

Also it depends on the type of expected occlusion. When random things can be behind the text, then going with the background (apparently interrupting the lines) is a good idea
![image](https://user-images.githubusercontent.com/1548516/98354321-d46eb700-2020-11eb-84cf-6ea75c56779a.png)

When the text is expected to be very contrasty, eg. both background and all data colors are light, it looks more pleasing to not have a shadow:
![image](https://user-images.githubusercontent.com/1548516/98354452-0718af80-2021-11eb-838a-9dc6a9811c13.png)

![image](https://user-images.githubusercontent.com/1548516/98354891-a9389780-2021-11eb-8a2c-7616bc686d32.png)

![image](https://user-images.githubusercontent.com/1548516/98355207-2106c200-2022-11eb-93b0-0184627eb8fb.png)

The NYT is also mostly refraining from shadows if not needed: 
![image](https://user-images.githubusercontent.com/1548516/98355675-c6ba3100-2022-11eb-8c7b-e0fd47ce40f7.png)

Fun thing, the amount of shadow may vary with importance, eg. "National average" has a wider one:
![image](https://user-images.githubusercontent.com/1548516/98354691-5c54c100-2021-11eb-8f5c-a649c61a36b0.png)

**Comment by monfera** [comment]
Tableau and many other analytical tools don't bother with borders, and it may not be because of weaker concern for accessibility; it may be because of the perceptual tradeoffs the borders introduce on otherwise predictably (knowably) colored backgrounds. I think such borders for every and all sunburst and treemap cells should be avoided, though they might be OK on non-leaf layers of multilayer treemaps.

![image](https://user-images.githubusercontent.com/1548516/98356957-befb8c00-2024-11eb-8e2b-cee3311a6719.png)

**Comment by miukimiu** [comment]
@monfera, I think this option **shouldn't be enabled** by default for most of the charts. 

Normaly you would user borders in cases like the one you shared:

![98354691-5c54c100-2021-11eb-8f5c-a649c61a36b0](https://user-images.githubusercontent.com/2750668/98359455-2d8a1b00-2020-11eb-8e2c-736cbcdf951a.png)

It's also very common in **maps**. Basically, because the label can overlap multiple background colors. 

<img width="883" alt="Screenshot 2020-11-06 at 11 21 22" src="https://user-images.githubusercontent.com/2750668/98379180-ebbc9d00-203e-11eb-8138-d20e64ce816f.png">


And I agree with you:

> I think such borders for every and all sunburst and treemap cells should be avoided, though they might be OK on non-leaf layers of multilayer treemaps.

Also, I agree that:

> capping (limiting) the border width in terms of screen pixel or EM/REM would be better than scaling. 

I think a thinner border works better.

> Tableau and many other analytical tools don't bother with borders

I think tools like Tableau **enforces good practices**. For example, in Tableau in a bar chart, if the label doesn't fit the bar, it doesn't appear. And it changes the color for black or white according to the bar color.

![Picture1-5](https://user-images.githubusercontent.com/2750668/98376387-4e139e80-203b-11eb-87cf-f966886fdd46.png)

Or in google charts, if the label doesn't have space inside the bar it goes outside and the label color adapts:
![vertical_bar_label_full](https://user-images.githubusercontent.com/2750668/98376419-59ff6080-203b-11eb-9004-08a03dc292ed.png)


Looking at what is being implemented on https://github.com/elastic/kibana/pull/81776. And because the label disappears when it doesn't fit the bar. I don't see any benefits of having a border around the label. I think the label should have the `textInverted` option enabled by default. 

<img width="1010" alt="97699418-623f2500-1aaa-11eb-8a90-6e9655450cc2" src="https://user-images.githubusercontent.com/2750668/98377750-186fb500-203d-11eb-8649-7cf4bf47f56b.png">
<img width="1011" alt="97699499-8b5fb580-1aaa-11eb-89af-7af0a2cba534" src="https://user-images.githubusercontent.com/2750668/98377761-1b6aa580-203d-11eb-9c92-cf45f950b8a3.png">


### When should we use the border? 

IMO **we should use when we know that a label is going to overlap multiple background colors**. For what is being implemented in Lens, and because the label doesn't overlap multiple backgrounds, I don't see why we would have borders. Just the `textInverted` option would improve the a11y.

### Should we improve the defaults for value labels border size?

I think yes. The Elastic Charts should have this functionality. But tools like Lens should enforce best practices and just use this type of label when is really necessary. In cases when labels can overlap multiple background colors.

> we should probably transfer that issue to the EUI repo

I' can work on the EUI theme to improve the border.

**Comment by monfera** [comment]
Thanks @miukimiu for your amazing and thoughtful writeup, incl. going after dataviz practices in the wild. I'm onboard with your conclusions and how they're structured

**Comment by wylieconlon** [comment]
Following up on this since we just got a [user asking about it](https://discuss.elastic.co/t/text-color-of-values-on-chart-bar-chart/274319). Seems like we've made a decision to reduce the default border, but it's not implemented yet?

**Comment by miukimiu** [comment]
Thanks, @wylieconlon. I'll prioritize this in EUI.
